src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -105,14 +110,94 @@ public static class FrozenDictionary
                 // the Equals/GetHashCode methods to be devirtualized and possibly inlined.
                 if (ReferenceEquals(comparer, EqualityComparer<TKey>.Default))
                 {
-                    // In the specific case of Int32 keys, we can optimize further to reduce memory consumption by using
-                    // the underlying FrozenHashtable's Int32 index as the keys themselves, avoiding the need to store the
-                    // same keys yet again.
-                    return typeof(TKey) == typeof(int) ?
-                        (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source) :
-                        new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
+#if NET7_0_OR_GREATER
+                    static FrozenDictionary<TKey, TValue> PickIntegerDictionary<TInt>(Dictionary<TKey, TValue> source)
+                        where TInt : struct, IBinaryInteger<TInt>
+                    {
+                        var k = (TInt[])(object)source.Keys.ToArray();
+                        var v = source.Values.ToArray();
+
+                        Array.Sort(k, v);
+
+                        var min = k[0];
+                        var max = k[k.Length - 1];
+                        var range = ulong.CreateTruncating(max - min);
+
+                        if (range == (ulong)k.Length - 1)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new ClosedRangeIntegerFrozenDictionary<TInt, TValue>(k, v);
+                        }
+                        else if (k.Length <= MaxItemsInSmallIntegerDictionary)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new SmallIntegerFrozenDictionary<TInt, TValue>(k, v);
+                        }
+                        else if (typeof(TInt) == typeof(int))
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source);
+                        }
+
+                        return new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
```

Comment:
Nit: above we're else-if-ing with bodies that always return so the elses aren't necessary, but then here this isn't in an else.  Personal style preference would be to use else for all of them or for none of them so that we're consistent in the method.

Reply:
Done.